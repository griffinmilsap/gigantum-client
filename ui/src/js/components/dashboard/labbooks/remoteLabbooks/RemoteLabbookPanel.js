// vendor
import React, { Component } from 'react';
import uuidv4 from 'uuid/v4';
import Highlighter from 'react-highlight-words';
import classNames from 'classnames';
import Moment from 'moment';
// muations
import ImportRemoteLabbookMutation from 'Mutations/ImportRemoteLabbookMutation';
import BuildImageMutation from 'Mutations/container/BuildImageMutation';
// store
import store from 'JS/redux/store';
import { setWarningMessage, setMultiInfoMessage } from 'JS/redux/actions/footer';
// queries
import UserIdentity from 'JS/Auth/UserIdentity';
// components
import LoginPrompt from 'Components/shared/modals/LoginPrompt';
import Loader from 'Components/common/Loader';
// assets
import './RemoteLabbookPanel.scss';
// config
import config from 'JS/config';

export default class RemoteLabbookPanel extends Component {
  constructor(props) {
    super(props);

    this.state = {
      labbookName: props.edge.node.name,
      owner: props.edge.node.owner,
      isImporting: false,
      showLoginPrompt: false,
    };
    this._importingState = this._importingState.bind(this);
    this._clearState = this._clearState.bind(this);
    this._closeLoginPromptModal = this._closeLoginPromptModal.bind(this);
    this._handleDelete = this._handleDelete.bind(this);
  }

  /**
    * @param {object} edge
    * validates user's session and then triggers toggleDeleteModal which passes parameters to the DeleteLabbook component
  */
  _handleDelete(edge) {
    if (localStorage.getItem('username') !== edge.node.owner) {
      setWarningMessage('You can only delete remote Projects that you have created.');
    } else {
      UserIdentity.getUserIdentity().then((response) => {
        if (navigator.onLine) {
          if (response.data) {
            if (response.data.userIdentity.isSessionValid) {
              this.props.toggleDeleteModal({
                remoteId: edge.node.id, remoteOwner: edge.node.owner, remoteLabbookName: edge.node.name, existsLocally: this.props.existsLocally,
              });
            } else {
              this.props.auth.renewToken(true, () => {
                this.setState({ showLoginPrompt: true });
              }, () => {
                this.props.toggleDeleteModal({
                  remoteId: edge.node.id, remoteOwner: edge.node.owner, remoteLabbookName: edge.node.name, existsLocally: this.props.existsLocally,
                });
              });
            }
          }
        } else {
          this.setState({ showLoginPrompt: true });
        }
      });
    }
  }

  /**
    * @param {}
    * fires when user identity returns invalid session
    * prompts user to revalidate their session
  */
  _closeLoginPromptModal() {
    this.setState({
      showLoginPrompt: false,
    });
  }

  /**
    *  @param {}
    *  changes state of isImporting to false
  */
  _clearState = () => {
    if (document.getElementById('dashboard__cover')) {
      document.getElementById('dashboard__cover').classList.add('hidden');
    }
    this.setState({
      isImporting: false,
    });
  }

  /**
    *  @param {}
    *  changes state of isImporting to true
  */
  _importingState = () => {
    document.getElementById('modal__cover').classList.remove('hidden');
    document.getElementById('loader').classList.remove('hidden');
    this.setState({
      isImporting: true,
    });
  }

  /**
    *  @param {owner, labbookName}
    *  imports labbook from remote url, builds the image, and redirects to imported labbook
    *  @return {}
  */
 importLabbook = (owner, labbookName) => {
   const self = this;
   const id = uuidv4();
   const remote = `https://repo.${config.domain}/${owner}/${labbookName}`;

   UserIdentity.getUserIdentity().then((response) => {
     if (navigator.onLine) {
       if (response.data) {
         if (response.data.userIdentity.isSessionValid) {
           this._importingState();
           setMultiInfoMessage(id, 'Importing Project please wait', false, false);
           const successCall = () => {
             this._clearState();

             setMultiInfoMessage(id, `Successfully imported remote Project ${labbookName}`, true, false);


             BuildImageMutation(
               owner,
               labbookName,
               false,
               (response, error) => {
                 if (error) {
                   console.error(error);
                   setMultiInfoMessage(id, `ERROR: Failed to build ${labbookName}`, null, true, error);
                 }
               },
             );

             self.props.history.replace(`/projects/${owner}/${labbookName}`);
           };
           const failureCall = (error) => {
             this._clearState();
             setMultiInfoMessage(id, 'ERROR: Could not import remote Project', null, true, error);
           };
           self.setState({ isImporting: true });

           ImportRemoteLabbookMutation(
             owner,
             labbookName,
             remote,
             successCall,
             failureCall,
             (response, error) => {
               if (error) {
                 failurecall(error);
               }
               self.setState({ isImporting: false });
             },
           );
         } else {
           this.props.auth.renewToken(true, () => {
             this.setState({ showLoginPrompt: true });
           }, () => {
             this.importLabbook(owner, labbookName);
           });
         }
       }
     } else {
       this.setState({ showLoginPrompt: true });
     }
   });
 }

 render() {
   const { props, state } = this;
   const { edge } = props;

   const descriptionCss = classNames({
     'RemoteLabbooks__row RemoteLabbooks__row--text': true,
     blur: state.isImporting,
   });
   const deleteCSS = classNames({
     RemoteLabbooks__icon: true,
     'Tooltip-data Tooltip-data--small': localStorage.getItem('username') !== edge.node.owner,
     'RemoteLabbooks__icon--delete': localStorage.getItem('username') === edge.node.owner,
     'RemoteLabbooks__icon--delete-disabled': localStorage.getItem('username') !== edge.node.owner,
   });

   return (
     <div
       key={edge.node.name}
       className="Card Card--300 column-4-span-3 flex flex--column justify--space-between"
     >

       <div className="RemoteLabbooks__row RemoteLabbooks__row--icon">
         {
          this.props.existsLocally
            ? (
              <button
                className="RemoteLabbooks__icon RemoteLabbooks__icon--cloud Tooltip-data Tooltip-data--small"
                data-tooltip="Project exists locally"
                disabled
              >
              Imported
              </button>
            )
            : (
              <button
                disabled={state.isImporting}
                className="RemoteLabbooks__icon RemoteLabbooks__icon--cloud-download"
                onClick={() => this.importLabbook(edge.node.owner, edge.node.name)}
              >
              Import
              </button>
            )
        }
         <button
           className={deleteCSS}
           data-tooltip={localStorage.getItem('username') !== edge.node.owner ? 'Can only delete remote projects created by you' : ''}
           disabled={this.state.isImporting || localStorage.getItem('username') !== edge.node.owner}
           onClick={() => this._handleDelete(edge)}
         >
           Delete
         </button>

       </div>

       <div className={descriptionCss}>

         <div className="RemoteLabbooks__row RemoteLabbooks__row--title">
           <h6 className="RemoteLabbooks__panel-title">
             <Highlighter
               highlightClassName="LocalLabbooks__highlighted"
               searchWords={[store.getState().labbookListing.filterText]}
               autoEscape={false}
               caseSensitive={false}
               textToHighlight={edge.node.name}
             />
           </h6>

         </div>

         <p className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--owner">{edge.node.owner}</p>
         <p className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--owner">{`Created on ${Moment(edge.node.creationDateUtc).format('MM/DD/YY')}`}</p>
         <p className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--owner">{`Modified ${Moment(edge.node.modifiedDateUtc).fromNow()}`}</p>

         <p className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--description">
           {
            (edge.node.description && edge.node.description.length)
              ? (
                <Highlighter
                  highlightClassName="LocalLabbooks__highlighted"
                  searchWords={[store.getState().labbookListing.filterText]}
                  autoEscape={false}
                  caseSensitive={false}
                  textToHighlight={edge.node.description}
                />
              )
              : (
                <span className="RemoteDatasets__description--blank">
              No description provided
                </span>
              )
           }
         </p>
       </div>

       { !(edge.node.visibility === 'local')
          && (
          <div
            data-tooltip={`${edge.node.visibility}`}
            className={`Tooltip-Listing RemoteLabbooks__${edge.node.visibility}   Tooltip-data Tooltip-data--small`}
          />
          )
       }

       {
          state.isImporting
          && (
          <div className="RemoteLabbooks__loader">
            <Loader />
          </div>
          )
        }

       {
          state.showLoginPrompt
          && <LoginPrompt closeModal={this._closeLoginPromptModal} />
        }
     </div>);
 }
}
