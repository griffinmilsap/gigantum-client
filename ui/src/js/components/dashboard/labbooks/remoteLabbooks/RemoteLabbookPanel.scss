@import '~Styles/_imports.scss';


.RemoteLabbooks__row Ã¥{
  &--icon {
    height: 68px;
    padding: 20px 20px 10px;

    @include flex(flex-end, row);
  }

  &--text {
    flex: 1;

    height: 100px;
    padding: 10px 20px;

    text-align: left;
  }

  &--title{
     @include flex(flex-start, row);
  }
}

.RemoteLabbooks__paragraph{
  color: $romanSilver;

  font-size: 14px;

  &--owner {
    color: $jet;
    font-size: 12px;
    height: 14px;
    margin: 2px;

    line-height: 14px;
  }

  &--description{
    display: -webkit-box;
    overflow: hidden;

    max-height: 130px;

    white-space: pre-line;
    text-overflow: ellipsis;
    word-break: normal;

    line-height: 20px;

    -o-text-overflow: ellipsis;
    -webkit-line-clamp: 5;
    line-clamp: 5;
    box-orient: vertical;
  }
}

.RemoteLabbooks__private{
  align-self: flex-end;

  width: 24px;
  height: 24px;
  padding: 20px;

  background: url(~Images/icons/lock-grey.svg) no-repeat center center;
  background-size: 18px;

}

.RemoteLabbooks__public{
  align-self: flex-end;

  width: 24px;
  height: 24px;
  padding: 20px;

  background: url(~Images/icons/earth-grey.svg) no-repeat center center;
  background-size: 18px;
}


.RemoteLabbooks__loader {
  position: absolute;
  top: 50%;
  left: 50%;

  transform: scale(0.5);
}

/*****************************
  RemoteLabbooks cloud
******************************/
.RemoteLabbooks__icon{
  width: 20px;
  height: 20px;
  margin-left: 5px;
  padding: 0;

  border: none;
}

.RemoteLabbooks__icon--cloud-download{
  background: url('~Images/icons/download.svg') no-repeat 7px center;
  background-size: 20px;
  color: $romanSilver;
  width: 125px;
  height: 26px;
  border-radius: 0;
  padding: 2px 0 2px 21px;
  border: 1px solid $romanSilver;

  &:hover,
   :focus{
    background: url('~Images/icons/download-azure.svg') no-repeat 7px center;
    background-size: 20px;
    width: 125px;
    height: 26px;
    border-radius: 0;
    padding: 2px 0 2px 21px;
    border: 1px solid $azure;
    color: $azure;
  }

  &:disabled{
    background: $background-color url('~Images/icons/download.svg') no-repeat 7px center;
    background-size: 20px;
    width: 125px;
    height: 26px;
    border-radius: 0;
    padding: 2px 0 2px 21px;
    border: 1px solid $romanSilver;
    color: $romanSilver;
  }
}

.RemoteLabbooks__icon--cloud{
  background: url('~Images/icons/download-check.svg') no-repeat 7px center;
  background-size: 20px;
  color: $romanSilver;
  height: 26px;
  width: 125px;
  border-radius: 0;
  padding: 2px 0 2px 21px;
  border: 1px solid $romanSilver;

  &:hover,
   :focus {
    background: $background-color url('~Images/icons/download-check.svg') no-repeat 7px center;
    background-size: 20px;
    height: 26px;
    width: 125px;
    border-radius: 0;
    padding: 2px 0 2px 21px;
    border: 1px solid $romanSilver;
  }

  &:disabled{
    background: $background-color url('~Images/icons/download-check.svg') no-repeat 7px center;
    background-size: 20px;
    height: 26px;
    width: 125px;
    border-radius: 0;
    padding: 2px 0 2px 21px;
    border: 1px solid $romanSilver;
  }
}

.RemoteLabbooks__icon--delete{
  background: url('~Images/icons/trash.svg') no-repeat 7px center;
  background-size: 20px;
  color: $romanSilver;
  height: 26px;
  width: 125px;
  border-radius: 0;
  padding: 2px 0 2px 21px;
  border: 1px solid $romanSilver;

  &:hover,
   :focus {
    background: url('~Images/icons/trash-azure.svg') no-repeat 7px center;
    background-size: 20px;
    height: 26px;
    width: 125px;
    border-radius: 0;
    padding: 2px 0 2px 21px;
    border: 1px solid $azure;
    color: $azure;
  }
}

.RemoteLabbooks__icon--delete-disabled,
.RemoteLabbooks__icon--delete-disabled:hover,
.RemoteLabbooks__icon--delete-disabled:focus,
.RemoteLabbooks__icon--delete-disabled:disabled{
  cursor: auto;
  background: $background-color url('~Images/icons/trash.svg') no-repeat 7px center;
  background-size: 20px;
  height: 26px;
  width: 125px;
  border-radius: 0;
  padding: 2px 0 2px 21px;
  border: 1px solid $romanSilver;
}

.RemoteDatasets__description--blank{
  font-style: italic;
}
