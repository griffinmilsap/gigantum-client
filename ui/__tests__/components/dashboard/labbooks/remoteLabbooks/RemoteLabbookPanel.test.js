// vendor
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import relayTestingUtils from '@gigantum/relay-testing-utils';
// Data
// history
import history from 'JS/history';
// components
import RemoteLabbookPanel from 'Components/dashboard/labbooks/remoteLabbooks/RemoteLabbookPanel';
import json from './__relaydata__/RemoteLabbooks.json';

const variables = { first: 20 };

const fixtures = {
  remoteLabbooks: json.data.labbookList.remoteLabbooks,
  toggleDeleteModal: () => {},
  labbookListId: json.data.labbookList.remoteLabbooks.id,
  className: 'LocalLabbooks__panel',
  edge: json.data.labbookList.remoteLabbooks.edges[0],
  histor: history,
  existsLocally: json.data.labbookList.remoteLabbooks.edges[0].node.isLocal,
  auth: {
    login: () => {
    },
  },
};


test('Test RemoteLabbooks rendering', () => {
  const localLabbooks = renderer.create(
    relayTestingUtils.relayWrap(
      <RemoteLabbookPanel {...fixtures} />, {}, json.data,
    ),
  );

  const tree = localLabbooks.toJSON();

  expect(tree).toMatchSnapshot();
});


export default variables;
