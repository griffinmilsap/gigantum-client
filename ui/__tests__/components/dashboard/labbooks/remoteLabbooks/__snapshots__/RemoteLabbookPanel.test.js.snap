// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test RemoteLabbooks rendering 1`] = `
<div
  className="Card Card--225 column-4-span-3 flex flex--column justify--space-between"
  data-selenium-id="RemoteLabbookPanel"
>
  <div
    className="RemoteLabbooks__row RemoteLabbooks__row--icon"
  >
    <div
      className="Tooltip-Listing RemoteLabbooks__private Tooltip-data Tooltip-data--small"
      data-tooltip="private"
    />
    <button
      className="Btn__dashboard Btn--action Btn__dashboard--cloud-download"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Import
    </button>
    <button
      className="Btn__dashboard Btn--action Tooltip-data Tooltip-data--wide Btn__dashboard--delete-disabled"
      data-tooltip="Only owners and admins can delete a remote Project"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
  </div>
  <div
    className="RemoteLabbooks__row RemoteLabbooks__row--text"
  >
    <div
      className="RemoteLabbooks__row RemoteLabbooks__row--title"
    >
      <h5
        className="RemoteLabbooks__panel-title"
      >
        <span>
          <span
            className=""
          >
            xd
          </span>
        </span>
      </h5>
    </div>
    <p
      className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--owner"
    >
      cbutler
    </p>
    <p
      className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--metadata"
    >
      <span
        className="bold"
      >
        Created:
      </span>
       
      Tue Jul 09 2019
    </p>
    <p
      className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--metadata"
    >
      <span
        className="bold"
      >
        Modified:
      </span>
       
      Tue Jul 09 2019
    </p>
    <p
      className="RemoteLabbooks__paragraph RemoteLabbooks__paragraph--description"
    >
      No description provided
    </p>
  </div>
</div>
`;
